design/apps.go

```diff
@@ -0,0 +1,260 @@
+	a.Description(`statistics and quotas for an enviromnent or deployment`)
+	a.Attribute("cpucores", envStatCores)
+	a.Attribute("memory", envStatMemory)
+	a.Attribute("pods", envStatPods)
+})
+
+var envStatCores = a.Type("EnvStatCores", func() {
+	a.Description(`CPU core stats`)
+	a.Attribute("used", d.Integer)
+	a.Attribute("quota", d.Integer)
+})
+
+var envStatMemory = a.Type("EnvStatMemory", func() {
+	a.Description(`memory stats`)
+	a.Attribute("used", d.Integer)
+	a.Attribute("quota", d.Integer)
+	a.Attribute("units", d.String)
+})
+
+var envStatPods = a.Type("EnvStatPods", func() {
+	a.Description(`pod stats`)
+	a.Attribute("starting", d.Integer)
+	a.Attribute("running", d.Integer)
+	a.Attribute("stopping", d.Integer)
+	a.Attribute("quota", d.Integer)
+})
+
+var timedIntTuple = a.Type("TimedIntTuple", func() {
+	a.Description("a set of time and integer values")
+	a.Attribute("time", d.Integer)
+	a.Attribute("value", d.Integer)
+})
+
+var simpleDeploymentStatSeries = a.Type("SimpleDeploymentStatSeries", func() {
+	a.Description("pod stat series")
+	a.Attribute("start", d.Integer)
+	a.Attribute("end", d.Integer)
+	a.Attribute("memory", a.ArrayOf(timedIntTuple))
+	a.Attribute("cores", a.ArrayOf(timedIntTuple))
+})
+
+var simpleSpaceSingle = JSONSingle(
+	"SimpleSpace", "Holds a single response to a space request",
+	simpleSpace,
+	nil)
+
+var simpleAppSingle = JSONSingle(
+	"SimpleApplication", "Holds a single response to a space/application request",
+	simpleApp,
+	nil)
+
+var simpleEnvironmentSingle = JSONSingle(
+	"SimpleEnvironment", "Holds a single response to a space/environment request",
+	simpleEnvironment,
+	nil)
+
+var simpleEnvironmentMultiple = JSONList(
+	"SimpleEnvironment", "Holds a response to a space/environment request",
+	simpleEnvironment,
+	nil,
+	nil)
+
+var simplePod = a.Type("SimplePod", func() {
+	a.Description("wrapper for a kubernetes Pod")
+	a.Attribute("pod", d.Any)
+})
+
+var simplePodMultiple = JSONList(
+	"SimplePod", "Holds a list of pods",
+	simplePod,
+	nil,
+	nil)
+
+var simpleDeploymentSingle = JSONSingle(
+	"SimpleDeployment", "Holds a single response to a space/application/deployment request",
+	simpleDeployment,
+	nil)
+
+var simpleDeploymentStatSeriesSingle = JSONSingle(
+	"SimpleDeploymentStatSeries", "HOlds a response to a stat series query",
+	simpleDeploymentStatSeries,
+	nil)
+
+var simpleEnvironmentStatSingle = JSONSingle(
+	"EnvStats", "Holds a single response to a pipeline/stats request",
+	envStats,
+	nil)
+
+var _ = a.Resource("apps", func() {
+	a.BasePath("/apps")
+
+	// An auth token is required to call the auth API to get an OpenShift auth token.
+	a.Security("jwt")
+
+	a.Action("showSpace", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID"),
+		)
+		a.Description("list applications in a space")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+		})
+		a.Response(d.OK, simpleSpaceSingle)
+		a.Response(d.NotFound, JSONAPIErrors)
```

Comment:
do you want to restrict the access to this endpoint ? if so, you should also return the `Unauthorized` response, and include auth verification in the controller methods.

Reply:
also, you should return an `InternalServerError` as well (we never know..)

Reply:
All endpoints in this API require auth (because I need auth to get the openshift token), but the current code doesn't validate access to spaces, because none of the rest of WIT does, unless the specific API modifies state.  This new API only does this in one place (pod scaling), and I assumed OpenShift would perform auth at that point.

Line 134 in design/apps.go ensures all endpoints require auth.  Is it a better style (it seems more error-prone) to define this for individual endpoints? 

I have added Unauthorized and InternalServerError to the list of responses on every endpoint.